acpi/src/lib.rs

```diff
@@ -95,189 +100,173 @@ pub enum AcpiError {
-        handler: H,
-        rsdp_mapping: PhysicalMapping<H, Rsdp>,
-    ) -> Result<AcpiTables<H>, AcpiError> {
+    ///
+    /// ### Safety: Caller must ensure that the provided mapping is a fully validated RSDP.
+    pub unsafe fn from_validated_rsdp(handler: H, rsdp_mapping: PhysicalMapping<H, Rsdp>) -> AcpiResult<Self> {
         let revision = rsdp_mapping.revision();
 
         if revision == 0 {
             /*
              * We're running on ACPI Version 1.0. We should use the 32-bit RSDT address.
              */
-            let rsdt_address = rsdp_mapping.rsdt_address();
-            unsafe { Self::from_rsdt(handler, revision, rsdt_address as usize) }
+
+            // ### Safety: Addresses from a validated `RSDP` are also guaranteed to be valid.
+            let rsdt_mapping: PhysicalMapping<H, SdtHeader> = unsafe {
+                handler.map_physical_region::<SdtHeader>(
+                    rsdp_mapping.rsdt_address() as usize,
+                    core::mem::size_of::<SdtHeader>(),
+                )
+            };
+            rsdt_mapping.validate(Signature::RSDT)?;
+
+            Ok(Self { mapping: rsdt_mapping, revision, handler })
         } else {
             /*
              * We're running on ACPI Version 2.0+. We should use the 64-bit XSDT address, truncated
              * to 32 bits on x86.
              */
-            let xsdt_address = rsdp_mapping.xsdt_address();
-            unsafe { Self::from_rsdt(handler, revision, xsdt_address as usize) }
+
+            // ### Safety: Addresses from a validated `RSDP` are also guaranteed to be valid.
+            let xsdt_mapping: PhysicalMapping<H, SdtHeader> = unsafe {
+                handler.map_physical_region::<SdtHeader>(
+                    rsdp_mapping.xsdt_address() as usize,
+                    core::mem::size_of::<SdtHeader>(),
+                )
+            };
+            xsdt_mapping.validate(Signature::XSDT)?;
+
+            Ok(Self { mapping: xsdt_mapping, revision, handler })
         }
     }
 
-    /// Create an `AcpiTables` if you have the physical address of the RSDT. This is useful, for example, if your chosen
-    /// bootloader reads the RSDP and passes you the address of the RSDT. You also need to supply the correct ACPI
-    /// revision - if `0`, a RSDT is expected, while a `XSDT` is expected for greater revisions.
-    pub unsafe fn from_rsdt(handler: H, revision: u8, rsdt_address: usize) -> Result<AcpiTables<H>, AcpiError> {
-        let mut result = AcpiTables { revision, sdts: BTreeMap::new(), dsdt: None, ssdts: Vec::new(), handler };
-
-        let header = sdt::peek_at_sdt_header(&result.handler, rsdt_address);
-        let mapping =
-            unsafe { result.handler.map_physical_region::<SdtHeader>(rsdt_address, header.length as usize) };
+    /// Searches through the ACPI table hears and attempts to locate the table with a matching `T::SIGNATURE`.
+    pub fn find_table<T: AcpiTable>(&self) -> AcpiResult<PhysicalMapping<H, T>> {
```

Comment:
Not sure where we want the boundary to be but I think this still needs to be `unsafe` - the problem here is that you can supply an incorrect `T` that then allows too much memory to be accessed, resulting in UB. In fact, a better way of marking this contract (i.e. the type being appropriate for a given table) would be to make the `AcpiTable` trait unsafe, and document this requirement there. What do you think?

Reply:
I agree with the latter solution—`unsafe trait AcpiTable` + documentation outlining this behaviour would be most semantic. Resolved!